internal/jimm/bootstrap/bootstrap_job.go

```diff
@@ -0,0 +1,245 @@
+// bootstrapping a controller and adding it to JIMM.
+func (b *bootstrapManager) BootstrapJob(
+	p JobParams,
+	executor BootstrapExecutor,
+	user *openfga.User,
+) jobtracker.JobFunc {
+	const bootstrapLockTTL = 40 * time.Minute
+
+	return func(ctx context.Context) error {
+		jobId, ok := jobtracker.JobIdFromContext(ctx)
+		if !ok {
+			return fmt.Errorf("failed to get job ID from context")
+		}
+
+		zapctx.Debug(
+			ctx,
+			"starting bootstrap job",
+			zap.String("job-id", jobId.String()),
+			zap.String("controller-name", p.ControllerName),
+		)
+
+		if err := b.store.LockBootstrap(ctx, bootstrapLockTTL); err != nil {
+			zapctx.Error(
+				ctx,
+				"failed to acquire bootstrap lock",
+				zap.String("job-id", jobId.String()),
+				zap.String("controller-name", p.ControllerName),
+				zap.Error(err),
+			)
+			return fmt.Errorf("failed to acquire bootstrap lock: %w", err)
+		}
+
+		defer func() {
+			if err := b.store.UnlockBootstrap(ctx); err != nil {
+				zapctx.Error(
+					ctx,
+					"failed to unlock bootstrap lock",
+					zap.String("job-id", jobId.String()),
+					zap.String("controller-name", p.ControllerName),
+					zap.Error(err),
+				)
+			}
+		}()
+
+		err := b.store.GetController(ctx, &dbmodel.Controller{Name: p.ControllerName})
+		if err == nil {
+			return errors.E(errors.CodeAlreadyExists, fmt.Errorf("controller %q already exists", p.ControllerName))
+		}
+		if errors.ErrorCode(err) != errors.CodeNotFound {
+			return fmt.Errorf("failed to check if controller exists: %w", err)
+		}
+
+		binary, err := b.binaryStore.Get(
+			ctx,
+			jujuclistore.JujuBinarySpec{
+				Version: p.CLIVersion,
+				Os:      p.CLIVersion,
+				Arch:    p.CLIArch,
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("failed to get Juju binary: %w", err)
+		}
+		defer binary.Done()
+
+		if err := b.runBootstrap(ctx, p, jobId, executor, binary, user); err != nil {
+			return fmt.Errorf("run bootstrap failed: %w", err)
+		}
+		return nil
+	}
+}
+
+// runBootstrap wraps the logic of running a controller bootstrap for JIMM into
+// a self-contained function. It is expected, and only expected to be run from within
+// the [bootstrapManager.BootstrapJob].
+func (b *bootstrapManager) runBootstrap(
+	ctx context.Context,
+	p JobParams,
+	jobId uuid.UUID,
+	executor BootstrapExecutor,
+	binary *jujuclistore.Binary,
+	user *openfga.User,
+) error {
+	outputCh, clientStore, cleanup, err := executor.RunWrapper(
+		ctx,
+		binary.FullPath,
+		p.JujuDataDir,
+		jujucommands.BootstrapCmdParams{
+			CloudNameAndRegion:   p.CloudNameAndRegion,
+			ControllerName:       p.ControllerName,
+			AgentVersion:         p.AgentVersion,
+			BootstrapTimeout:     p.BootstrapTimeout,
+			LoginTokenRefreshURL: p.LoginTokenRefreshURL,
+			PersonalCloud:        p.PersonalCloud,
+			CloudCred:            p.CloudCred,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to run bootstrap command: %w", err)
+	}
+	defer cleanup()
+
+	for output := range outputCh {
```

Comment:
so here we are waiting for the bootstrap process to finish, right?

Reply:
Yup.

Reply:
what if request timeout expires and we're not done bootstrapping?

Reply:
then it'll return on line 186 and cleanup will run.